### ISSUE TITLE
Provide a 'mut' slang

### ISSUE THREAD

**Issue by masak** [masak/007#118]
Let's define an *index path* as a variable (a *root*), followed by one or more
index operations:

    x                   # index path
    y[7]                # index path
    z.foo[0]["bar"]     # index all the way
    w()                 # nope
    n + 1               # nah

The idea is to capture the notion of "lvalue" in 007: something that we are
able to modify, if only by updating the value in the root.

We could imagine a convenience function to help do such updates. We can't
write `a[0][1][2] = "foo";` in 007, because updates only happen on variables.
But we could write something like

    a = mutate(a, sub (context) {
        context.set(0, sub (value) {
            return mutate(value, sub (context) {
                context.set(1, sub (value) {
                    return mutate(value, sub (context) {
                        context.set(2, "foo");
                    });
                });
            });
        });
    });

Basically, each indexing operation translates to a call to `context.set`, until
we bottom out in a primitive value.

This is straightforward, and not too hard to implement, but definitely no fun
to write. So let's introduce a slang to make it more palatable:

    a = mut {
        _[0][1][2] = "foo";
    };

This desugars to the above nested function expression. The `_` variable is
forcibly introduced into the `mut` block context without any explicit
declaration. Inside a `mut` block, the left-hand side of assignments can
be any index path of `_`, suppressing the normal requirement in 007 that it
be a simple identifier.

As a special case, we could have

    n = mut {
        _ = _ + 1;
    };

Desugar to just `n = n + 1`. Simply because that's what the user would expect.

Allowing several assignments inside a `mut` shouldn't pose any problems. The
update would still be transactional, that is, the whole structure gets replaced
all at once. The final result should be exactly as if mutation of index paths
were possible in the main language. Assignments should be evaluated in order.
For example:

    a = [0, 1, 2];
    a = mut {
        _[2] = _[0];
        _[0] = "zow!";
    };

Should assign `["zow!", 1, 0]` to `a`, and not `["zow!", 1, "zow!"]`.

It should also be possible to call `.push`, `.pop` (and optionally `.shift` and
`.unshift`) on arrays.

    state = mut {
        my disk = _[oldpile].pop();
        _[newpile].push(disk);
    };

Even with these operations, assignments happen in order. For example, an index
operations doesn't commute with a `.shift` operation.

There must be an assignment operator (`=`) just to the left of the `mut` macro
keyword; otherwise, the macro dies in agony.

Because this assignment is itself an expression, there's also a method to
control the value of this expression:

    n = mut {
        _.yield(_);
        _ = _ + 1;
    };

(This example emulates the `prefix:<++>` operator found in many languages.) The
`.yield` call can appear anywhere in the block, but it's illegal to have more
than one. If no `.yield` call is present, the resulting value that gets
assigned back to the variable is also the value of the whole assignment
expression.

In fact, we could easily provide both `prefix:<++>` and `postfix:<++>` and
their corresponsing `--` forms as macros that make use of the `mut` slang. So
those would work seamlessly not only on variables, but on any index path.

We could take this further and supply `+=`, `*=` et al., with the usual
semantics. This would probably tempt us to toy around with metaoperators in
007, since one would expect a new infix operator `⊚` to get a corresponding
`⊚=` form automatically.

I'm not too worried about that. But I am a bit concerned that good ol'
assignment (`=`) will seem a bit underpowered next to all the sugary macro-y
operators, and that people will attempt to assign to index paths outside of a
`mut`, get the compilation error, and curse 007 for being inconsistent.

**Comment by masak** [comment]
Here, for what it's worth, is a diff showing how the hanoi.007 example would look under the `mut` slang:

```diff
diff --git a/examples/hanoi.007 b/examples/hanoi.007
index 0a7aaae..cc1885f 100644
--- a/examples/hanoi.007
+++ b/examples/hanoi.007
@@ -13,54 +13,6 @@ my state = [
     []
 ];
 
-sub upto(n) {
-    my c = -1;
-    sub counter(irrelevant_I) {
-        c = c + 1;
-        return c;
-    }
-
-    return map(counter, [1] xx n);
-}
-
-sub pop(n) {
-    sub pop_from_nth_pile(nth) {
-        my pile = state[nth];
-        if n != nth {
-            return pile;
-        }
-
-        sub pile_item(i) { return pile[i] }
-
-        return map(pile_item, upto(elems(pile) - 1));
-    }
-
-    my L = elems(state[n]) - 1;
-    my disk = state[n][L];
-    state = map(pop_from_nth_pile, upto(3));
-    return disk;
-}
-
-sub push(disk, n) {
-    sub push_to_nth_pile(nth) {
-        my pile = state[nth];
-        if n != nth {
-            return pile;
-        }
-
-        sub pile_item(i) {
-            if i == elems(pile) {
-                return disk;
-            }
-            return pile[i];
-        }
-
-        return map(pile_item, upto(elems(pile) + 1));
-    }
-
-    state = map(push_to_nth_pile, upto(3));
-}
-
 sub tw(n, L) {
     my e = elems(state[n]);
     if L >= e {
@@ -70,7 +22,7 @@ sub tw(n, L) {
 }
 
 sub show() {
-    for reversed(upto(5)) -> L {
+    for [4, 3, 2, 1, 0] -> L {
         say(tw(0, L) ~ " " ~ tw(1, L) ~ " " ~ tw(2, L));
     }
 }
@@ -79,8 +31,10 @@ sub move(diskname, from, to) {
     say("");
     say("Moving " ~ diskname ~ " from pile " ~ str(from + 1) ~ " to pile " ~ str(to + 1) ~ "...");
     say("");
-    my disk = pop(from);
-    push(disk, to);
+    state = mut {
+        my disk = _[from].pop();
+        _[to].push(disk);
+    };
     show();
 }
 ```

**Comment by masak** [comment]
Though this issue is very likely going to lose to #119, I'll note that React's [`update`](http://facebook.github.io/react/docs/update.html) looks a lot saner than what I came up with re the `mut` DSL.

**Comment by masak** [comment]
Closing because we just closed #119.

**Comment by masak** [comment]
Just want to drive-by comment that [immer](https://github.com/mweststrate/immer) is very close to what I had in mind for this functionality.
